File: ios/RNStepCounter.m
#import "RNStepCounter.h"
#import "SOMotionDetecter.h"

#import <CoreMotion/CoreMotion.h>
#import <React/RCTBridge.h>
#import <React/RCTEventDispatcher.h>

@interface RNStepCounter ()
@property (nonatomic, readonly) CMPedometer *pedometer;
@end

@implementation RNStepCounter
+ (BOOL)requiresMainQueueSetup {
    return YES;
}

@synthesize bridge = _bridge;
@synthesize callableJSModules = _callableJSModules;

RCT_EXPORT_MODULE();

- (NSArray<NSString *> *)supportedEvents {
    return @[@"StepCounter.stepCounterUpdate", @"StepCounter.stepDetected"];
}

RCT_EXPORT_METHOD(isStepCountingSupported:(RCTPromiseResolveBlock)resolve
                                   reject:(RCTPromiseRejectBlock)reject) {
    resolve(@{@"granted": @(self.authorizationStatus),
            @"supported": @([CMPedometer isStepCountingAvailable]),
                 @"pace": @([CMPedometer isPaceAvailable]),
              @"cadence": @([CMPedometer isCadenceAvailable]),
             @"distance": @([CMPedometer isDistanceAvailable]),
           @"floorCount": @([CMPedometer isFloorCountingAvailable]),
    });
}

RCT_EXPORT_METHOD(queryStepCounterDataBetweenDates:(NSDate *)startDate
                                           endDate:(NSDate *)endDate
                                           handler:(RCTResponseSenderBlock)handler) {
    [self.pedometer queryPedometerDataFromDate:startDate
                                        toDate:endDate
      withHandler:^(CMPedometerData *pedometerData, NSError *error) {
          handler(@[error.description?:[NSNull null], [self dictionaryFromPedometerData:pedometerData]]);
      }];
}

RCT_EXPORT_METHOD(startStepCounterUpdate:(NSDate *)date) {
    [self.pedometer startPedometerUpdatesFromDate:date?:[NSDate date]
                                      withHandler:^(CMPedometerData *pedometerData, NSError *error) {
        if (pedometerData) {
            [self sendEventWithName:@"StepCounter.stepCounterUpdate"
             body:[self dictionaryFromPedometerData:pedometerData]];
        }
    }];
}

- (NSDictionary *)dictionaryFromPedometerData:(CMPedometerData *)data {
    static NSDateFormatter *formatter;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        formatter = [[NSDateFormatter alloc] init];
        formatter.dateFormat = @"yyyy-MM-dd'T'HH:mm:ss.SSSZZZZZ";
        formatter.locale = [NSLocale localeWithLocaleIdentifier:@"en_US_POSIX"];
        formatter.timeZone = [NSTimeZone timeZoneWithName:@"UTC"];
    });
    return @{
        @"counterType": @"CMPedometer",
          @"startDate": [formatter stringFromDate:data.startDate]?:[NSNull null],
            @"endDate": [formatter stringFromDate:data.endDate]?:[NSNull null],
              @"steps": data.numberOfSteps?:[NSNull null],
           @"distance": data.distance?:[NSNull null],
     @"floorsAscended": data.floorsAscended?:[NSNull null],
    @"floorsDescended": data.floorsDescended?:[NSNull null],
    };
}

RCT_EXPORT_METHOD(stopStepCounterUpdate) {
    [self.pedometer stopPedometerUpdates];
    [[SOMotionDetecter sharedInstance] stopDetection];
}

RCT_EXPORT_METHOD(startStepsDetection) {
    [[SOMotionDetecter sharedInstance]
      startDetectionWithUpdateBlock:^(NSError *error) {
        if(error) {
            return;
        } else {
            [self sendEventWithName:@"StepCounter.stepDetected"
                               body:@true];
        }
    }];
}

- (BOOL)authorizationStatus {
#if __IPHONE_OS_VERSION_MAX_ALLOWED >= 110000
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wpartial-availability"
    CMAuthorizationStatus status = [CMPedometer authorizationStatus];
    return status == CMAuthorizationStatusAuthorized;
#pragma clang diagnostic pop
#endif
}

#pragma mark - Private

- (instancetype)init {
    self = [super init];
    if (self == nil) {
        return nil;
    }
    _pedometer = [[CMPedometer alloc]init];
    return self;
}

@end

File: ios/RNStepCounter.swift
import CoreMotion
import Foundation
import React

@objc(RNStepCounter)
class RNStepCounter: RCTEventEmitter {
    private let stepCounter: CMPedometer = .init()

    var numberOfSteps: Int! = 0

    override func supportedEvents() -> [String]! {
        return ["stepCounterUpdate"]
    }

    override static func requiresMainQueueSetup() -> Bool {
        return true
    }

    @objc(isStepCountingSupported:withRejecter:)
    func isStepCountingSupported(resolve: RCTPromiseResolveBlock, reject _: RCTPromiseRejectBlock) {
        if CMPedometer.isStepCountingAvailable() {
            resolve(true)
        } else {
            resolve(false)
        }
    }

    @objc(startStepCounterUpdate)
    func startStepCounterUpdate(_ date: NSDate?) {
        guard authorizationStatus() else {
            fatalError("cant' auth")
        }
        stepCounter.startUpdates(from: date ?? NSDate(), withHandler: { data, error in
            guard let pedometerData = data, error == nil else {
                print("There was an error getting the data: \(String(describing: error))")
                return
            }
            if pedometerData != nil {
                let pedDataSteps = pedometerData.numberOfSteps.intValue
                DispatchQueue.main.async {
                    if self.numberOfSteps != pedDataSteps {
                        print("Steps: \(pedometerData)")
                        self.numberOfSteps = pedDataSteps
                        self.sendEvent(
                            withName: "stepCounterUpdate",
                            body: self.dictionaryFromPedometerData(pedometerData!)
                        )
                    }
                }
            }
        })
    }

    func dictionaryFromPedometerData(_ data: CMPedometerData) -> [String: Any] {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSZZZZZ"
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone(abbreviation: "UTC")
        return [
            "startDate": formatter.string(from: data.startDate) ?? NSNull(),
            "endDate": formatter.string(from: data.endDate) ?? NSNull(),
            "steps": data.numberOfSteps ?? NSNull(),
            "distance": data.distance ?? NSNull(),
            "floorsAscended": data.floorsAscended ?? NSNull(),
            "floorsDescended": data.floorsDescended ?? NSNull()
        ]
    }

    func authorizationStatus() -> Bool {
        var stepCounterAuth = false
        if #available(iOS 11.0, *) {
            switch CMPedometer.authorizationStatus() {
            case .authorized:
                // 완전히 허용한 경우
                stepCounterAuth = true
            case .restricted:
                // 제한적인 권한으로 승인한 경우
                stepCounterAuth = true
            case .notDetermined:
                // 일회성으로 허용하는 경우
                stepCounterAuth = true
            case .denied:
            // 완전히 거절한 경우
            @unknown default:
                // 사용이 불가능한 기종이나 컨디션인 경우
                break
            }
        } else {
            // Fallback on earlier versions (iOS 10 이하)
        }
        return stepCounterAuth
    }

    @objc(stopStepCounterUpdate)
    func stopStepCounterUpdate() {
        stepCounter.stopUpdates()
        if #available(iOS 10.0, *) {
            stepCounter.stopEventUpdates()
        } else if numberOfSteps != 0 {
            // Fallback on earlier versions (iOS 9 이하)
            numberOfSteps = 0
            sendEvent(withName: "stepCounterUpdate", body: ["steps": numberOfSteps])
        }
    }

    override init() {
        pedometer = CMPedometer()
    }
}
